ext/mbstring/libmbfl/mbfl/mbfilter.c

```diff
@@ -83,2575 +83,1527 @@
+			return MAX((long)length - (char_offset << 1), 0);
+		} else if (encoding->flag & (MBFL_ENCTYPE_WCS4BE | MBFL_ENCTYPE_WCS4LE)) {
+			return MAX((long)length - (char_offset << 2), 0);
+		} else if (encoding == &mbfl_encoding_utf8) {
+			/* We can easily count backwards from the end of a UTF-8 string
+			 * As we back up, decrement `char_offset` when we find a byte which is the first byte in a character
+			 * It's easy to tell, since any byte which starts with bits 10 is a continuation byte and is not
+			 * the first byte in a character */
+			unsigned char *p = start + length;
+			while (char_offset > 0 && --p > start) {
+				unsigned char c = *p;
+				if (c < 0x80) {
+					char_offset--;
+				} else if ((c & 0xC0) != 0x80) {
+					char_offset--;
+				}
+			}
+			return p - start;
+		} else if (encoding == &mbfl_encoding_utf16le) {
+			/* We can also easily count backwards from the end of a UTF-16 string */
+			length &= -2; /* If there is a 'junk byte' at the end of the string, leave it out */
+			unsigned char *p = start + length;
+			while (char_offset > 0 && (p -= 2) > start) {
+				unsigned char c = *p;
+				if (c < 0xD8 || c >= 0xE0) {
+					char_offset--;
+				} else if ((c & 0xFC) == 0xD8) {
+					char_offset--;
+				}
+			}
+			return p - start;
+		} else {
+			/* If we don't have any efficient way to iterate backwards from the end
+			 * of the string, the only sane thing to do is to iterate through
+			 * the string _twice_. First, go through one time and find its length
+			 * in characters. Then, go through a second time and stop when we reach
+			 * the requested number of characters before the end.
+			 *
+			 * (If `char_offset` is a small number, we could also keep track of the
+			 * last `char_offset` character positions passed by when iterating through
+			 * the string. This would avoid the need to iterate over the string twice.
+			 * But it's not good for the general case, since `char_offset` can be
+			 * arbitrarily large.) */
+			long char_len = byte_offset_to_char_offset(encoding, start, length, length);
+			if (char_offset >= char_len) {
+				return 0;
+			}
+			return char_offset_to_byte_offset(encoding, start, length, char_len - char_offset);
+		}
+	} else if (char_offset >= length) {
+		return length;
+	}
+
+	/* `char_offset` is positive; we are searching from the beginning of the string.
+	 * The easiest case is when each character takes a constant number of bytes */
+	if (encoding->flag & MBFL_ENCTYPE_SBCS) {
+		return MIN(char_offset, length);
+	} else if (encoding->flag & (MBFL_ENCTYPE_WCS2BE | MBFL_ENCTYPE_WCS2LE)) {
+		return MIN(char_offset * 2, length);
+	} else if (encoding->flag & (MBFL_ENCTYPE_WCS4BE | MBFL_ENCTYPE_WCS4LE)) {
+		return MIN(char_offset * 4, length);
+	} else if (encoding->mblen_table != NULL) {
+		const unsigned char *mbtab = encoding->mblen_table;
+		unsigned char *p = start, *e = p + length;
+		while (p < e && char_offset--) {
+			p += mbtab[*p];
+		}
+		return MIN(p, e) - start;
```

Comment:
```suggestion
		return MIN(p - start, length);
```
For symmetry with other cases?